Feature/41 exception hendler

stishchenko opened the pull request.

stishchenko committed:
#41 @ExceptionHandler
pom -> add starter to validation.
ErrorInfo -> class to store info about error in AppExceptionHandler.
AppExceptionHandler -> realisation of AOP with @RestControllerAdvice to catch validation and other exceptions.
ProductController -> remove try-catch blocks.
ProductService -> change product fields checking.
ProductDto -> describe constrains with validation annotations, replace no args constructor by lombok annotation.

stishchenko committed:
#41 change updateProduct test in ProductServiceTest

stishchenko committed:
try resolve prise test

NckGrs commented on src/main/java/team/asd/controller/ProductController.java:
```diff
@@ -21,37 +22,21 @@ public ProductController(@Autowired ProductService productService) {
-        } catch (ValidationException e) {
-            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
-        }
+        return new ResponseEntity<>(ConverterUtil.convertProductToDto(productService.readById(id)), HttpStatus.OK);
```
You don’t need to wrap the DTO in a ResponseEntity, since you use the @RestController annotation, I would also recommend returning the DTO and not the Object in the method.

stishchenko committed:
#41 Changing of controller returning.
ProductController -> replace ResponseEntity with ProductDto or Integer(for deleting).
ProductDto -> remove version.
ConverterUtil -> remove version.
Product -> change LocalDateTime with Date; remove unused constructors.

NckGrs commented on src/main/java/team/asd/exception/AppExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorInfo catchUnexpectedException(Exception e) {
+        return new ErrorInfo(HttpStatus.BAD_REQUEST.value(), "Unexpected error: " + e.getMessage());
```
When an unexpected error occurs, the error message should not contain the message from the exception, as this may indicate a vulnerability in the code.

stishchenko committed:
AppExceptionHandler -> remove exception message from catchUnexpectedException().

rdovgan committed:
Formatted code to avoid changes in PR

rdovgan committed:
Merge remote-tracking branch 'origin/master' into feature/41_Exception_hendler

# Conflicts:
#	pom.xml
#	src/main/java/team/asd/controller/ProductController.java
#	src/main/java/team/asd/controller/TestController.java
#	src/main/java/team/asd/dto/ProductDto.java
#	src/main/java/team/asd/entity/Product.java
#	src/main/java/team/asd/service/ProductService.java
#	src/main/java/team/asd/util/ConverterUtil.java
#	src/test/java/team/asd/service/ProductServiceTest.java

rdovgan committed:
Formatted code to avoid changes in PR
